client/DESIGN.md

```diff
@@ -0,0 +1,380 @@
+# Treode Client Library Interface
+
+The two major components of the Treode client library interface are the [Cache](#Cache) and the [Transaction](#Transaction).  
+
+Caches store key-value pairs retrieved from the server, as well as the read and value timestamps associated with each pair that Treode provides.  Caches are responsible for maintaining a connection to the DB server.
+
+Transactions allows users to locally perform a sequence of database operations and, when finished, optimistically commit the result to the server by writing through the cache.
+
+<a name="Cache"></a>
+# Cache
+
+We provide an overview of the Cache class below, including its primary internal data structure, its public methods, and the form of its HTTP requests and responses.
+
+- Parameters: 
+    1. server: string (required)
+    2. port: string (port=80)
+    3. max\_age: int (max\_age=None)
+    4. no\_cache: bool (no\_cache=False)
+- Public Methods: 
+    1. read(table: string, key: string, max\_age=None, no\_cache=False): value (JSON)
+    2. write(table: string, op\_list: list): status (string), throws StaleException
+- Internal Objects: 
+    1. HTTP library for establishing/managing connection to server (in Python, urllib3)
+    2. map for maintaining cache
+- Exceptions: 
+    1. StaleException
+    2. StaleException(Value_time: int)
+
+## Cache Structure Description
+
+The cache internally maintains (1) a connection to the DB server and (2) a map of cache entries.  
+
+### Database Connection
+
+The cache should utilize an HTTP Request/Response library (e.g., in Python, urllib3) to establish a connection with the DB server and read and write values through its course of use.  We describe in the HTTP Message Formatting section the expected format for DB requests and responses.
+
+[Here are examples](https://urllib3.readthedocs.org/en/latest/) of using the urllib3 library in Python to open and use a connection.
+
+### Map of Cache Entries
+
+We propose two possible designs for implementing the cache; the preferable of the two designs depends largely on the existing libraries available in the application language.
+
+#### Two-Level Map Implementation
+
+The cache is a two-level map of the following type:
+
+    table : key -> value-time -> value : read-time
```

Comment:
Sometimes this uses x:y to indicate a tuple, and other times (x, y). I think the confusion stems from using "x:y" as a proxy for a tuple in Python dict. Let's consistently use (x, y) to indicate a tuple, and make a note about how that translates to keys in a Python dict.